### ISSUE TITLE
Clarification about verifying NFM Ansatz for Transformer

### ISSUE THREAD

**Issue by 123epsilon** [aradha/agop_feature_learning#1]
Hi I just wanted some clarification on the logic when verifying the NFM ansatz for transformers as implemented in `/ansatz_verification/transformers/language_models/verify_ansatz.py`

1) In `check_ansatz` you seem to be [applying dropout](https://github.com/aradha/agop_feature_learning/blob/69128656dfd9a7b99cd069dd3a56f1ba9eb6de11/ansatz_verification/transformers/language_models/verify_ansatz.py#L248) to the input to our original transformer before running it through all the layers - why is this done? 
```python
 with torch.no_grad():
            n, t = x.shape
            pos = torch.arange(0, t, dtype=torch.long, device='cuda')
            z = Q.drop(Q.wpe(pos) + Q.wte(x))

            for i in range(layer_idx):
                z = Q.h[i](z)
```

2) Directly after the above step, `z` is the output after running the input tensor through `layer_idx+1` full blocks of our transformer, but we then seem to arbitrarily [apply the layernorm ](https://github.com/aradha/agop_feature_learning/blob/69128656dfd9a7b99cd069dd3a56f1ba9eb6de11/ansatz_verification/transformers/language_models/verify_ansatz.py#L254) for the `layer_idx+1`th block corresponding to its `CausalSelfAttention` module- why is this done? From my understanding `z` should already be normalized (and in particular, has already passed through `ln_2`). 

```python
 with torch.no_grad():
            ln_z1 = Q.h[layer_idx].ln_1(z).unsqueeze(0)
            ln_z2 = deepcopy(ln_z1)
            ln_z3 = deepcopy(ln_z1)
            ln_z = deepcopy(ln_z1)
```

3) Also, since it seems like the goal is to evaluate the NFM for attention, why are we taking the output of each block after it has already passed through the MLP layers for the corresponding block - shouldn't we hook in and get it right after it passes through our attention block?

4) When we actually seek to [get the jacobian](https://github.com/aradha/agop_feature_learning/blob/69128656dfd9a7b99cd069dd3a56f1ba9eb6de11/ansatz_verification/transformers/language_models/verify_ansatz.py#L271) of the model we run the renormalized `ln_z1,...,ln_z` back through the network using the `newGPT` object. Why are we passing the outputs of the network from the block at `layer_idx` back through the same network? Shouldn't we be passing in tokens corresponding to the text that we care about (i.e. `x`) and then extract the gradients for the attention matrices we care about? 

```python
 J = get_jacobian(newGPT, [z.unsqueeze(0), ln_z1, ln_z2, ln_z3, ln_z],
                                 BATCH_SIZE=BATCH_SIZE,
                                 IDX=IDX,
                                 OUT_SIZE=OUT_SIZE,
                                 OUT_IDX=OUT_IDX)
```

It's possible that I'm misunderstanding the implementation somewhat but I'd love to apply this work to some other transformer analysis and I'd like to know how these quantities are actually computed in your paper - thank you!

**Comment by aradha** [comment]
Thanks for the questions - hope the following helps clarify!  At a high level, to compute the ansatz for the NFA, we need to compute the AGOP with respect to inputs to attention.  Importantly, this needs to be done separately for the key, query and value matrices since they operate on the input in different ways.  

(1) Our first goal is to get access to the input to attention.  To this end, we are matching the forward pass of our GPT2 model.  The forward pass of the model first computes a dropout after summing the token and positional embeddings (see line 200 of model.py): 
`x = self.transformer.drop(tok_emb + pos_emb)`

(2) For the same reason as in (1), we apply layer norm.  Namely, the input to attention in a GPT2 block first applies layernorm to the data (line 125 of model.py): 
`x = x + self.attn(self.ln_1(x))`

(3, 4) We want to get access to self.ln_1(x) in the line above since this is the input to attention (denoted as self.attn). So our grads should be with respect to these.  You could of course compute the AGOP with respect to the inputs themselves, but this will capture first layer features (token and positional embedding).

**Comment by 123epsilon** [comment]
Thank you for the clarification, that definitely helps but I'm still a little confused about the `get_jacobian` step, maybe I'm misunderstanding some of the `functorch` operators. 

From my understanding of the code we are:
1) grabbing the intermediate layer output `z` from `Block` `L-1` and then applying `ln_1` from `Block` `L` as this is the expected input to the next attention block
2) We take this normalized `ln_z` and then pass it through the network, computing the jacobian with respect to this input.

My confusion arises from the following:
1) I don't understand why we need to apply `ln_1`ourselves, since when we run it through the `Block` at layer `L` since this should be applied anyway through the forward pass, this is related to question 2 below
2) In `get_jacobian` we are running this input through `newGPT` whose forward pass runs the entire `GPT` model, I would expect that we would instead only run the hidden layers after the `Block` corresponding to `layer_idx`. That is, to use `jacrev` to compute the jacobian wrt to `ln_z` (or really, `z`) after running the latter half of the model
3) Why are there multiple inputs to `get_jacobian`? Especially since they are all copies of one another and don't seem to ever be modified I'm not sure why this is part of the implementation

Thank you again for taking the time to help me understand, I'm wondering whether my concept of the implementation is wrong or whether there are implementation details that just differ from my mental model

**Comment by aradha** [comment]
I think the confusion may be coming from the fact that newGPT is not the entire GPT model, but rather the model from a subset of the GPT model.  Namely, if we are interested in computing the ansatz at attention layer i, newGPT starts from layer from layer i onwards.  The subsetting happens in lines 224-225:
`T.h[layer_idx] = CustomBlock(T.h[layer_idx])
  T.h = T.h[layer_idx:]`

This is why, we needed to run the model up to that point (including dropout, layer norms, etc., to get the appropriate input into newGPT).  Does this clarify your questions 1. and 2.? 

The answer to your question 3. is subtle. The reason we need 5 inputs is the following.  The model does not apriori know that the operands of the key, query, value matrices are all the same.  So our ansatz implies that we need to treat them as different variables and compute AGOP with respect to each operand separately (these are the the three variables ln_z1, ln_z2, ln_z3).  The variables are copied so that the model doesn't accidentally tie them together in the gradient computation.  The variables z and ln_z going into the Jacobian are to handle the skip connections and multiple heads.  Namely, we do not want to include the skip connection gradients in our AGOP computation since the weights we want to capture are not involved in when adding the skip connection term.  Due to the nature of the skip connections in attention, we need to keep track z.  We keep track of ln_z to handle the AGOP for multiple heads.  We want to compute AGOP for each head's key, query and value separately since they all use different weight matrices.  We compute these AGOPs one at a time per head.  For these reasons, we wrote a CustomBlock and CustomAttention layer (lines 120 and 63).  

Hope this was helpful!
